### PR TITLE
Build the site when something is pushed to main, or on PRs

### LINKED ISSUE #7: Setup continuous integration to ensure that PRs work properly, provide artifacts for deployment

### PR DESCRIPTION
**Pull request by niloc132**
There is room for improvement here, this is meant to be a first draft
before we start merging other cleanup, to validate those merges.

Fixes #7

### CODE REVIEW

**Comment by niloc132** [comment]
(wrong issue commented on)

**Comment by niloc132** [comment]
Build summary, with downloadable generated JS https://github.com/niloc132/gwt-site-webapp/actions/runs/4338262581 (requires #11)

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      - uses: actions/checkout@v3
+      - name: Set up Java
+        uses: actions/setup-java@v3
+        with:
+          java-version: '11'
+          distribution: 'temurin'
+          cache: 'maven'
+
+      - name: Compile the webapp to JS
+        run: mvn --batch-mode -e verify
```
Comment (tbroyer) [inline .github/workflows/build.yml]: nit: add `--no-transfer-progress` (or `-ntp`) to make the logs less verbose and more readable

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+name: Build gwt-site-webapp
+on:
+  push:
+    branches: ['main']
```
Comment (tbroyer) [inline .github/workflows/build.yml]: This prevents forks (i.e. contributors) from running the workflow before making a pull request.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+          cache: 'maven'
+
+      - name: Compile the webapp to JS
+        run: mvn --batch-mode -e verify
+
+      - name: Upload the generated JS
+        uses: actions/upload-artifact@v3
+        with:
+          name: 'gwt-site-webapp'
+          path: target/gwt-site-webapp-2.0-SNAPSHOT/
```
Comment (tbroyer) [inline .github/workflows/build.yml]: Use a wildcard to not depend on the `<version>`?

File: pom.xml [hunk 4]
```diff
@@ -37,27 +36,22 @@
...
         </dependency>
         <dependency>
-            <groupId>com.google.gwt.site</groupId>
-            <artifactId>gwt-site</artifactId>
-            <version>${project.version}</version>
-            <type>zip</type>
-            <classifier>generated-site</classifier>
-         </dependency>
+            <groupId>com.google.gwt</groupId>
+            <artifactId>gwt-codeserver</artifactId>
```
Comment (tbroyer) [inline pom.xml]: Isn't content of gwt-codeserver already in gwt-dev for a while?

File: pom.xml [hunk 5]
```diff
@@ -6,21 +6,20 @@
...
-    <packaging>war</packaging>
+    <packaging>gwt-app</packaging>
 
     <properties>
         <maven.build.timestamp.format>yyyy-MM-dd</maven.build.timestamp.format>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 
-        <gwt.version>2.7.0</gwt.version>
-        <gwtquery.version>1.4.2</gwtquery.version>
+        <gwt.version>2.9.0</gwt.version>
```
Comment (tbroyer) [inline pom.xml]: Move to using the BOM?

File: pom.xml [hunk 6]
```diff
@@ -90,90 +73,51 @@
...
-                            <mode>htmlunit</mode>
-                            <gen>gen</gen>
-                            <extraJvmArgs>-Xss1024k -Xmx512M -XX:MaxPermSize=512m</extraJvmArgs>
-                            <saveSource>true</saveSource>
-                            <module>com.google.gwt.site.webapp.GWTProject</module>
-                            <!--
-                            Draft and pretty mode are turned on so that beginners can try out debugging
-                            in their browser with no setup.
-                            Please don't disable this; code size is less important.
-                            -->
```
Comment (tbroyer) [inline pom.xml]: Keep this comment? Otherwise someone will later want to change those values.

File: pom.xml [hunk 7]
```diff
@@ -90,90 +73,51 @@
...
-                        <id>copy-resources</id>
+                        <!-- Copy results necessary for sourcemaps to work to the packaging output -->
+                        <id>copy-sourcemap-resources</id>
                         <phase>prepare-package</phase>
                         <goals>
                             <goal>copy-resources</goal>
                         </goals>
                         <configuration>
-                            <outputDirectory>${project.build.directory}/www/gwtproject/</outputDirectory>
+                            <outputDirectory>${project.build.directory}/${project.artifactId}-${project.version}/</outputDirectory>
```
Comment (tbroyer) [inline pom.xml]:
```suggestion
                            <outputDirectory>${project.build.directory}/${project.build.finalName}/</outputDirectory>
```

File: pom.xml [hunk 8]
```diff
@@ -182,4 +126,65 @@
...
+                                <id>copy-www-resources</id>
+                                <phase>prepare-package</phase>
+                                <goals>
+                                    <goal>copy-resources</goal>
+                                </goals>
+                                <configuration>
+                                    <outputDirectory>${project.build.directory}/www/</outputDirectory>
+                                    <resources>
+                                        <resource>
+                                            <directory>${project.build.directory}/${project.artifactId}-${project.version}/</directory>
```
Comment (tbroyer) [inline pom.xml]:
```suggestion
                                            <directory>${project.build.directory}/${project.build.finalName}/</directory>
```

File: README.md [hunk 9]
```diff
@@ -17,36 +16,24 @@
 - To **build** and check the app, run
 
         $ cd [...]/gwt-site-webapp
-        $ mvn package
+        $ mvn verify
```
Comment (tbroyer) [inline README.md]: Shouldn't the `full-site` profile be mentioned somewhere in the README?

**Review by zbynek — APPROVED** [review]

File: pom.xml [hunk 10]
```diff
@@ -81,6 +81,11 @@
                     <compilerArgs>
                         <arg>-saveSource</arg>
                     </compilerArgs>
+                    <!--
+                    Draft and pretty mode are turned on so that beginners can try out debugging
```
Comment (zbynek) [inline pom.xml:85]: Would be better to rely on source maps and keep the code size reasonably small as it affects SEO. But this PR only adds the comment, so 👍

Comment (niloc132) [reply pom.xml:85]: This is actually an old comment that you removed in your PR, i'm just restoring it as the explanation for why we have (or had) this set
https://github.com/gwtproject/gwt-site-webapp/commit/27070768ae49c048535358a03c70a5b9d84ddf98#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L130-L134

I suspect that optimized js ends up being different enough that sometimes sourcemaps can't really "fix" where each line of code goes - like if a function is inlined, "stepping in" doesn't make sense any more. I'm not sure I agree with this perspective, but there are some easier ways to bring the size down for the moment if we're concerned about that.

**Pull request merged**